### PR TITLE
Core remove legacy patch

### PR DESCRIPTION
**Pull request by Berserker66**
Companion PR that should be merged at around the same time: https://github.com/ArchipelagoMW/z3randomizer/pull/4

## What is this fixing or adding?
removes a bunch of many-months old code that should no longer be needed.

## How was this tested?
barely, with a bit of LttP

### CODE REVIEW

File: worlds/alttp/Rom.py [hunk 1]
```diff
@@ -3,9 +3,9 @@
 import Utils
 from Patch import read_rom
 
-LTTPJPN10HASH = '03a63945398191337e896e5771f77173'
-RANDOMIZERBASEHASH = '9952c2a3ec1b421e408df0d20c8f0c7f'
-ROM_PLAYER_LIMIT = 255
+LTTPJPN10HASH: str = "03a63945398191337e896e5771f77173"
+RANDOMIZERBASEHASH: str = "9952c2a3ec1b421e408df0d20c8f0c7f"
+ROM_PLAYER_LIMIT: int = 255
```
Comment (black-sliver) [inline worlds/alttp/Rom.py:8]: I don't think that adding typing here helps :thinking: Typing deduction should be possible through the assignment

File: worlds/alttp/Rom.py [hunk 2]
```diff
@@ -123,29 +123,24 @@ def verify(buffer, expected: str = RANDOMIZERBASEHASH) -> bool:
         return expected == buffermd5.hexdigest()
 
     def patch_base_rom(self):
-        if os.path.isfile(local_path('basepatch.sfc')):
-            with open(local_path('basepatch.sfc'), 'rb') as stream:
+        if os.path.isfile(user_path('basepatch.sfc')):
+            with open(user_path('basepatch.sfc'), 'rb') as stream:
```
Comment (black-sliver) [inline worlds/alttp/Rom.py:127]: i don't even want to know how i missed that

File: Patch.py [hunk 3]
```diff
@@ -299,141 +179,3 @@ def read_rom(stream, strip_header=True) -> bytearray:
     if strip_header and len(buffer) % 0x400 == 0x200:
         return buffer[0x200:]
     return buffer
-
-
-if __name__ == "__main__":
```
Comment (black-sliver) [inline Patch.py:304]: do we really want to get rid of `python Patch.py path/to/patch` completely? I used that a couple of times during development. A stripped-down version is probably enough though.

File: Patch.py [hunk 4]
```diff
@@ -162,102 +154,11 @@ def patch(self, target: str):
             f.write(result)
 
 
-# legacy patch handling follows:
 GAME_ALTTP = "A Link to the Past"
 GAME_SM = "Super Metroid"
 GAME_SOE = "Secret of Evermore"
 GAME_SMZ3 = "SMZ3"
 GAME_DKC3 = "Donkey Kong Country 3"
```
Comment (black-sliver) [inline Patch.py:161]: We should try to get rid of those as well, but that's probably out of scope here, since that'd be SNIClient rework.

**Review by black-sliver — COMMENTED** [review]
Still have to test, but left a few comments as food for thought

File: worlds/alttp/Rom.py (hunk 1 above)
Comment (Berserker66) [reply worlds/alttp/Rom.py:8]: yeah, but it also doesn't hurt.

File: Patch.py (hunk 3 above)
Comment (Berserker66) [reply Patch.py:304]: A wrapper around APDeltaPatch could have some use, though a cmd flag for the client to behave as just a patcher could accomplish the same thing.

File: Patch.py (hunk 4 above)
Comment (Berserker66) [reply Patch.py:161]: agreed on both. So that's more towards @PoryGone 's scope right now.

File: Patch.py [hunk 5]
```diff
@@ -302,138 +191,6 @@ def read_rom(stream, strip_header=True) -> bytearray:
...
-                                print(f"File {future.result()} added to {os.path.split(updated_zip)[1]}")
-
-            except:
-                import traceback
-
-                traceback.print_exc()
-                input("Press enter to close.")
+    for file in sys.argv[1:]:
+        meta_data, result_file = create_rom_file(file)
+        print(f"Patch with meta-data {meta_data} was written to {result_file}")
```
Comment (black-sliver) [inline Patch.py:196]: Interestingly this does not work.

The first issue is that you have to `import worlds.AutoWorld` to get the patch handlers to register - that might be a general problem with create_rom_file, but SNIClient luckily imports the worlds first.
If you try to import worlds.AutoWorld from Patch.py you end up with a cyclic reference that leaves the local copy of AutoPatchRegister.file_endings empty.

The only suggestion i have right now is
1. move AutoPatchRegister, APContainer and APDeltaPatch to e.g. AutoPatch.py
2. `import worlds.AutoWorld; del AutoWorld` before `def create_rom_file`

Comment (Berserker66) [reply Patch.py:196]: maybe we add --patch_only or such to SNIClient?

**Review by black-sliver — COMMENTED** [review]
* Tested gen of basepatch.sfc
* Tested gen of  alttp, sm, smz3, soe, factorio.
* Tested patching of alttp, sm, smz3, soe (either Patch.py or SNIClient.py)
* Tested SNIClient.py patching and launching
* Tested Patch.py patching
* Tested LttpAjuster.py xxx.aplttp, LttpAdjuster.py xxx.sfc, LttpAdjuster gui
* Tested webhost upload, webhost download, webhost gen

it all worked.

**I did not test dkc3.**

Also see comments below.

Also maybe check if doc/* needs updating?

The only other thing I found is that LttpAdjuster does not validate the input file if it's a .sfc and breaks e.g. soe roms, it does not overwrite the original though.

File: Patch.py [hunk 6]
```diff
@@ -1,273 +1,30 @@
...
-import lzma
-import threading
-import concurrent.futures
-import zipfile
 import sys
-from typing import ClassVar, List, Tuple, Optional, Dict, Any, Union, BinaryIO
+from typing import Tuple, Optional, TypedDict
 
-import ModuleUpdate
-ModuleUpdate.update()
```
[the remark this answers is not in the document]
Comment (Berserker66) [reply Patch.py:16]: CommonClient does, but too late, mh.

**Comment by PoryGone** [comment]
> **I did not test dkc3.**

I can test this branch against DKC3 in an hour or so, if that is desired.

**Comment by Berserker66** [comment]
... I don't get the test failures locally. Time to dig I suppose.

**Comment by Berserker66** [comment]
Might help to update local branch to remote branch when I did the merge in github. SMW is hopefully working now.

**Review by black-sliver — APPROVED** [review]
Can be merged as soon as we get the OK for DKC3 and someone tests SMW

**Comment by Berserker66** [comment]
Tested SMW in local gen and webhost gen -> patch -> connect, result of that is the latest commit.

**Comment by PoryGone** [comment]
Tested DKC3 with local gen and webhost gen, patching with `Patch.py` and with `SNIClient.py`. Everything seems to behave as expected.

**Pull request merged**